src/main/java/spoon/reflect/visitor/OperatorHelper.java

```diff
@@ -219,4 +244,277 @@ public static OperatorAssociativity getOperatorAssociativity(UnaryOperatorKind o
+		if (!leftType.isPrimitive()
+			|| !rightType.isPrimitive()
+			|| leftType.equals(booleanType)
+			|| rightType.equals(booleanType)) {
+			return Optional.empty();
+		}
+
+		CtTypeReference<?> doubleType = typeFactory.DOUBLE_PRIMITIVE;
+		// If either operand is of type double, the other is converted to double.
+		if (leftType.equals(doubleType) || rightType.equals(doubleType)) {
+			return Optional.of(doubleType);
+		}
+
+		// Otherwise, if either operand is of type float, the other is converted to float.
+		CtTypeReference<?> floatType = typeFactory.FLOAT_PRIMITIVE;
+		if (leftType.equals(floatType) || rightType.equals(floatType)) {
+			return Optional.of(floatType);
+		}
+
+		// Otherwise, if either operand is of type long, the other is converted to long.
+		CtTypeReference<?> longType = typeFactory.LONG_PRIMITIVE;
+		if (leftType.equals(longType) || rightType.equals(longType)) {
+			return Optional.of(longType);
+		}
+
+		// Otherwise, both operands are converted to type int.
+		return Optional.of(typeFactory.INTEGER_PRIMITIVE);
+	}
+
+	/**
+	 * Get the promoted type of a binary operator, as defined by the Java Language Specification.
+	 *
+	 * @param operator the operator
+	 * @param left the left operand
+	 * @param right the right operand
+	 * @return the promoted type or {@link Optional#empty()} if promotion does not apply or the operation is invalid
+	 */
+	public static Optional<CtTypeReference<?>> getPromotedType(
+		BinaryOperatorKind operator,
+		CtExpression<?> left,
+		CtExpression<?> right
+	) {
+		TypeFactory typeFactory = left.getFactory().Type();
+		switch (operator) {
+			// logical operators
+			case AND:
+			case OR: {
+				CtTypeReference<?> booleanType = typeFactory.BOOLEAN_PRIMITIVE;
+				if (!left.getType().equals(booleanType) || !right.getType().equals(booleanType)) {
+					return Optional.empty();
+				}
+
+				return Optional.of(booleanType);
+			}
+			// shift operators are special:
+			case SL:
+			case SR:
+			case USR: {
+				// See: https://docs.oracle.com/javase/specs/jls/se11/html/jls-15.html#jls-15.19
+				// on each operand unary numeric promotion is performed
+				CtTypeReference<?> promotedLeft = unaryNumericPromotion(left).orElse(null);
+				CtTypeReference<?> promotedRight = unaryNumericPromotion(right).orElse(null);
+
+				if (promotedLeft == null || promotedRight == null) {
+					return Optional.empty();
+				}
+
+				// after promotion, both operands have to be an integral type:
+				if (!isIntegralType(promotedLeft) || !isIntegralType(promotedRight)) {
+					return Optional.empty();
+				}
+
+				// The type of the shift expression is the promoted type of the left-hand operand.
+				return Optional.of(promotedLeft);
+			}
+			case INSTANCEOF:
+				// See: https://docs.oracle.com/javase/specs/jls/se11/html/jls-15.html#jls-15.20.2
+				// Not implemented, because it is not necessary for the current use case.
+				throw new IllegalStateException("instanceof is not yet implemented");
```

Comment:
Same as with EQ, NE; I think this can just always be boolean?

Reply:
It seems like I have to document better what this method does. The method does **not** return the type of the expression (that would be unnecessary, because `CtBinaryOperator#getType` will return it).

Instead it will check if an expression can be applied to an operator (if not, it will return empty).

There are some types like `short` that are cast to another type (in this case `int`) before the operator is applied. The types seem to always be the same after promotion on both sides, so there is no distinction made.

Returning `boolean` for the instanceof operator, would result in the left and right being cast to `boolean` before applying the operator (which is obviously wrong).

Thinking about 
> types seem to always be the same after promotion on both sides

it seems like that would not apply to `instanceof`.

Reply:
The problem is that I have to check any context where `lhs instanceof rhs` is not a valid operation. According to the [JLS](https://docs.oracle.com/javase/specs/jls/se11/html/jls-15.html#jls-15.20.2), I would have to check
- that the lhs is a reference type or null type (should be easy to implement)
- that the rhs is [reifiable](https://docs.oracle.com/javase/specs/jls/se11/html/jls-4.html#jls-4.7) (sounds like a lot more work)